cmd/src/debug_kube.go

```diff
@@ -0,0 +1,340 @@
+		log.Printf("Archiving kubectl data for %d pods\n SRC_ENDPOINT: %v\n Context: %s Namespace: %v\n Output filename: %v", len(pods.Items), cfg.Endpoint, kubectx, namespace, base)
+		if verified, _ := verify("Do you want to start writing to an archive?"); !verified {
+			return nil
+		}
+
+		err = archiveKube(ctx, zw, *verbose, noConfigs, namespace, baseDir, pods)
+		if err != nil {
+			return cmderrors.ExitCode(1, err)
+		}
+		return nil
+	}
+
+	debugCommands = append(debugCommands, &command{
+		flagSet: flagSet,
+		handler: handler,
+		usageFunc: func() {
+			fmt.Println(usage)
+		},
+	})
+}
+
+type podList struct {
+	Items []struct {
+		Metadata struct {
+			Name string
+		}
+		Spec struct {
+			Containers []struct {
+				Name string
+			}
+		}
+	}
+}
+
+// Runs common kubectl functions and archive results to zip file
+func archiveKube(ctx context.Context, zw *zip.Writer, verbose, noConfigs bool, namespace, baseDir string, pods podList) error {
+	// Create a context with a cancel function that we call when returning
+	// from archiveKube. This ensures we close all pending go-routines when returning
+	// early because of an error.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// setup channel for slice of archive function outputs, as well as throttling semaphore
+	ch := make(chan *archiveFile)
+	g, ctx := errgroup.WithContext(ctx)
+	semaphore := semaphore.NewWeighted(8)
+
+	run := func(f func() error) {
+		g.Go(func() error {
+			if err := semaphore.Acquire(ctx, 1); err != nil {
+				return err
+			}
+			defer semaphore.Release(1)
+
+			return f()
+		})
+	}
+
+	// create goroutine to get pods
+	run(func() error {
+		ch <- getPods(ctx, namespace, baseDir)
+		return nil
+	})
+
+	// create goroutine to get kubectl events
+	run(func() error {
+		ch <- getEvents(ctx, namespace, baseDir)
+		return nil
+	})
+
+	// create goroutine to get persistent volumes
+	run(func() error {
+		ch <- getPV(ctx, namespace, baseDir)
+		return nil
+	})
+
+	// create goroutine to get persistent volumes claim
+	run(func() error {
+		ch <- getPVC(ctx, namespace, baseDir)
+		return nil
+	})
+
+	// start goroutine to run kubectl logs for each pod's container's
+	for _, pod := range pods.Items {
+		for _, container := range pod.Spec.Containers {
+			p := pod.Metadata.Name
+			c := container.Name
+			run(func() error {
+				ch <- getPodLog(ctx, p, c, namespace, baseDir)
+				return nil
+			})
+		}
+	}
+
+	// start goroutine to run kubectl logs --previous for each pod's container's
+	// won't write to zip on err, only passes bytes to channel if err not nil
+	for _, pod := range pods.Items {
+		for _, container := range pod.Spec.Containers {
+			p := pod.Metadata.Name
+			c := container.Name
+			run(func() error {
+				f := getPastPodLog(ctx, p, c, namespace, baseDir)
+				if f.err == nil {
+					ch <- f
+				} else if verbose {
+					fmt.Printf("Could not gather --previous pod logs for: %s \nExited with err: %s\n", p, f.err)
+				}
+				return nil
+			})
+		}
+	}
```

Comment:
This one looks a bit different, but can conform:

```suggestion
			run(func() *archiveFile {
				f := getPastPodLog(ctx, p, c, namespace, baseDir)
				if f.err != nil {
					fmt.Printf("Could not gather --previous pod logs for: %s \nExited with err: %s\n", p, f.err)
					return nil
				}
				return f
			})
		}
	}
```